docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,132 @@
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
+
+### Cons
+- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
+- Implicit error flow. Errors are thrown via GHC exceptions in `IO`. See exception handling below.
+
+## Exception handling
+
+### Exception type
+```haskell
+data CustomCliException where
+  CustomCliException
+    :: (Show error, Typeable error, Error error, HasCallStack)
+    => error -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+
+The purpose of `CustomCliException` is to represent explicitly thrown, structured errors that are meaningful to our application. 
+For other exceptions (e.g., runtime errors or unexpected failures), we rely on the `Nothing` case in our proposed top level
+handler below:
+
+### Top level handler 
+```haskell
+executeRio :: RIO () () -> IO ()
+executeRio r = do
+  runRIO () r
+    `catch` ( \e ->
+                case fromException e of
+                  Just custom@(CustomCliException{}) -> hPutStrLn stderr $ displayException custom
+                  Nothing -> hPutStrLn stderr $ displayException e
+            )
+```
+
+### Usage Example of CustomCliException
+
+https://github.com/IntersectMBO/cardano-cli/pull/1033
+
+
+### Exception Handling Mechanism: Pros & Cons
+
+#### Pros
+
+1. **Unified Exception Type**
+
+- Simplifies Top-Level Handling: All errors are caught as `CustomCliException`.
+- Consistent Reporting: Ensures all errors are formatted uniformly via `prettyError`.
+
+2. **CallStack Inclusion**
+
+- Embeds CallStack to trace error origins, aiding debugging.
+
+3. **Polymorphic Error Support**
+
+- Flexibility: Wraps any error type so long at the required instances exist.
+
+#### Cons
+1. **Type Erasure**
+
+- Loss of Specificity: Existential quantification erases concrete error types, preventing pattern-matching on specific errors. That may make specific error recovery logic harder to implement.
+
+# Decision
+- Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.
+- Top level monad: [RIO](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO). 
+
+# Consequences
+- This should dramatically improve our code's composability and remove many unnecessary error types. 
+- Readability concerning what errors can be thrown will be negatively impacted. However `ExceptT` already lies about what exceptions can be thrown because it is not limited to the error type stated in `ExceptT`'s type signature. In other words `IO` can implicitly throw other `Exception`s. 
```

Comment:
```suggestion
- Readability concerning what errors can be thrown will be negatively impacted. However, `ExceptT` already lies about what exceptions can be thrown because it is not limited to the error type stated in `ExceptT`'s type signature. In other words, `IO` can implicitly throw other `Exception`s. 
```